### DISCUSSION TITLE
Getting ${NPM_TOKEN} errors in 25.49.1

### DISCUSSION THREAD

**Discussion by protoz** [renovatebot/renovate#10602] [Request Help] [answered]
**How are you running Renovate?**

- [ ] WhiteSource Renovate hosted app on github.com
- [x] Self hosted

If using the hosted app, please skip to the next section.
Otherwise, if self-hosted, please complete the following:

Please select which platform you are using:

- [ ] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server
- [ ] Bitbucket Cloud (bitbucket.org)
- [ ] Bitbucket Server
- [ ] Gitea
- [ ] github.com
- [x] GitHub Enterprise Server
- [ ] gitlab.com
- [ ] GitLab self-hosted

Renovate version: 25.49.1

**Describe the bug**

We've been having issues with private modules and using the NPM token we are adding through our Concourse job.
```WARN: Failed to replace env in config: ${NPM_TOKEN} (repository=my-project/my-project-repo)```

Which looks to be a yarn error message if the token is not specified in the .npmrc file.

Since we are running the 25.x series we were under the impressions ${NPM_TOKEN} wasn't used and renovate did not look at the .npmrc file anymore.  

Our host config matches the example in documentation
```       
         { 
            hostType: "npm",
            matchHost: "registry.npmjs.org",
            token: process.env.NPMJS_TOKEN,
        },
```

We do have ${NPM_TOKEN} specified in our .npmrc which matches the documentation in this repository - [private-npm-modules.md](https://github.com/renovatebot/renovate/blob/main/docs/usage/private-npm-modules.md)

That documentation also mentions to not use ${NPM_TOKEN} which is a bit confusing.  Can we not use that environment variable at all in our repos even if it is only specified in the now unread .npmrc?

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
renovate debug
DEBUG: Setting npmrc (repository=my-project/my-project-repo) 
WARN: Failed to replace env in config: ${NPM_TOKEN} (repository=my-project/my-project-repo) 
DEBUG: Datasource unknown error (repository=my-project/my-project-repo) 
"datasource": "npm", 
"lookupName": "@hapi/boom", 
"err": { 
"message": "env-replace", 
"stack": "Error: env-replace\n at /usr/local/lib/node_modules/renovate/dist/datasource/npm/npmrc.js:31:19\n at String.replace (<anonymous>)\n at envReplace (/usr/local/lib/node_modules/renovate/dist/datasource/npm/npmrc.js:28:18)\n at Object.setNpmrc (/usr/local/lib/node_modules/renovate/dist/datasource/npm/npmrc.js:80:26)\n at Object.getReleases (/usr/local/lib/node_modules/renovate/dist/datasource/npm/releases.js:12:17)\n at fetchReleases (/usr/local/lib/node_modules/renovate/dist/datasource/index.js:224:36)\n at getRawReleases (/usr/local/lib/node_modules/renovate/dist/datasource/index.js:251:25)\n at Object.getPkgReleases (/usr/local/lib/node_modules/renovate/dist/datasource/index.js:267:35)\n at Object.lookupUpdates (/usr/local/lib/node_modules/renovate/dist/workers/repository/process/lookup/index.js:58:61)\n at fetchDepUpdates (/usr/local/lib/node_modules/renovate/dist/workers/repository/process/fetch.js:40:36)\n at async /usr/local/lib/node_modules/renovate/node_modules/p-map/index.js:57:22" 
} 
DEBUG: Failed to look up dependency @hapi/boom (repository=my-project/my-project-repo, packageFile=package.json, dependency=@hapi/boom) 
DEBUG: Found deprecationMessage (repository=my-project/my-project-repo, dependency=@hapi/joi) 
DEBUG: Dependency lookup failure: not found (repository=my-project/my-project-repo) 
"err": { 
"name": "HTTPError", 
"timings": { 
"start": 1624482873993, 
"socket": 1624482873993, 
"lookup": 1624482874031, 
"connect": 1624482874042, 
"secureConnect": 1624482874069, 
"upload": 1624482874069, 
"response": 1624482874415, 
"end": 1624482874415, 
"phases": { 
"wait": 0, 
"dns": 38, 
"tcp": 11, 
"tls": 27, 
"request": 0, 
"firstByte": 346, 
"download": 0, 
"total": 422 
} 
}, 
"message": "Response code 404 (Not Found)", 
"stack": "HTTPError: Response code 404 (Not Found)\n at Request.<anonymous> (/usr/local/lib/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:117:42)\n at runMicrotasks (<anonymous>)\n at processTicksAndRejections (node:internal/process/task_queues:96:5)", 
"options": { 
"headers": { 
"user-agent": "https://github.com/renovatebot/renovate", 
"authorization": "***********", 
"accept": "application/json", 
"accept-encoding": "gzip, deflate, br" 
}, 
"url": "https://registry.npmjs.org/@my-company%2Fprivate-plugin", 
"username": "", 
"password": "", 
"method": "GET", 
"http2": false 
}, 
"response": { 
"statusCode": 404, 
"statusMessage": "Not Found", 
"body": {"error": "Not found"}, 
"headers": { 
"date": "Wed, 23 Jun 2021 21:14:34 GMT", 
"content-type": "application/json", 
"content-length": "21", 
"connection": "close", 
"cf-ray": "6640ba0afb332a09-ORD", 
"www-authenticate": "Basic, Bearer", 
"cf-cache-status": "DYNAMIC", 
"cf-request-id": "0adc529adb00002a099e8ad000000001", 
"expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"", 
"vary": "Accept-Encoding", 
"server": "cloudflare" 
}, 
"httpVersion": "1.1" 
} 
}, 
"packageName": “@my-company/private-plugin"

```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [x] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

...

**Comment by rarkins** [comment, marked as answer]
> Since we are running the 25.x series we were under the impressions ${NPM_TOKEN} wasn't used and renovate did not look at the .npmrc file anymore.

Renovate does not look at `~/.npmrc`, but files in the repo are another matter.

> We do have ${NPM_TOKEN} specified in our .npmrc which matches the documentation in this repository

No, that does not match the documentation. The documentation is talking about the `npmrc` field in config combined with `npmToken`.

What is your full server config? e.g. do you run with `exposeAllEnv`?

**Reply by protoz** [reply]
Here is the full config, sorry I was out of the office and couldn't grab the config til this morning.  Yes we are setting `exposeAllEnv: true`

```
module.exports = {
    hostRules: [
        {
            matchHost: "https://gcr.io/my-company-containers/",
            username: "_json_key",
            password: process.env.SERVICE_ACCOUNT_CREDENTIALS,
        },
        {
            matchHost: "artifactory.my-company.com",
            username: process.env.ARTIFACTORY_USER,
            password: process.env.ARTIFACTORY_PASSWORD,
        },
        {
            hostType: "npm",
            matchHost: "registry.npmjs.org",
            token: process.env.NPMJS_TOKEN,
        },
        {
            hostType: "npm",
            matchHost: "registry.yarnpkg.com",
            token: process.env.NPMJS_TOKEN,
        },
        {
            hostType: "npm",
            matchHost: "https://registry.yarnpkg.com/@my-company/",
            token: process.env.NPMJS_TOKEN,
        },
    ],
    exposeAllEnv: true,
    allowedPostUpgradeCommands: [
        "^npm install$",
        "^npm install --global npm --no-save$"
    ],
};
```

**Reply by rarkins** [reply]
If you add NPM_TOKEN to your env (seems like it should be same as NPMJS_TOKEN) then does that resolve it?

**Reply by busches** [reply]
We'll give it a whirl, but that's specifically called out in the docs as to not do that: https://github.com/renovatebot/renovate/blob/main/docs/usage/private-npm-modules.md#L82

**Reply by protoz** [reply]
Adding NPM_TOKEN back has fixed the issue.  Some clarity in that document would be appreciated since it contradicts itself which is why we moved NPM_TOKEN to NPMJS_TOKEN in the first place.  Marking resolved.

**Reply by viceice** [reply]
The `NPM_TOKEN` hint was outdated and removed from docs.
